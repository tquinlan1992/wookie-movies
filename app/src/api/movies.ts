import axios from "axios";

// const sampleMovie = {
//     backdrop: "https://wookie.codesubmit.io/static/backdrops/d6822b7b-48bb-4b78-ad5e-9ba04c517ec8.jpg"
// cast: ['Christian Bale', 'Heath Ledger', 'Aaron Eckhart']
// classification: "13+"
// director: "Christopher Nolan"
// genres: ['Action', 'Crime', 'Drama']
// id: "d6822b7b-48bb-4b78-ad5e-9ba04c517ec8"
// imdb_rating: 9
// length: "2h 32min"
// overview: "Batman raises the stakes in his war on crime. With the help of Lt. Jim Gordon and District Attorney Harvey Dent, Batman sets out to dismantle the remaining criminal organizations that plague the streets. The partnership proves to be effective, but they soon find themselves prey to a reign of chaos unleashed by a rising criminal mastermind known to the terrified citizens of Gotham as the Joker."
// poster: "https://wookie.codesubmit.io/static/posters/d6822b7b-48bb-4b78-ad5e-9ba04c517ec8.jpg"
// released_on: "2008-07-16T00:00:00"
// slug: "the-dark-knight-2008"
//
// }

export interface Movie {
  backdrop: string;
  cast: string[];
  classification: string;
  director: string;
  genres: string[];
  id: string;
  imdb_rating: 9;
  length: string;
  overview: string;
  poster: string;
  released_on: string;
  slug: string;
  title: string;
}

export const getMovies = async (search: string) => {
  const moviesResponse = await axios.get<{ movies: Movie[] }>(
    "https://wookie.codesubmit.io/movies",
    {
      headers: { Authorization: "Bearer Wookie2021" },
      params: { q: search },
    }
  );
  return moviesResponse.data.movies;
};
