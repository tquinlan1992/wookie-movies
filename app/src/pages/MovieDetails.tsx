const MovieDetails: React.FC = () => <h3>MovieDetails</h3>;

export default MovieDetails;
