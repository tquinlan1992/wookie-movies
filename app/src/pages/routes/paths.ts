export const paths = {
  HOME: "/",
  MOVIE_DETAILS: `/movies/:movieTitle/:movieId`,
};

export const toPaths = {
  HOME: () => paths.HOME,
  MOVIE_DETAILS: ({ id, title }: { id: string; title: string }) =>
    `/movies/${title}/${id}`,
};
