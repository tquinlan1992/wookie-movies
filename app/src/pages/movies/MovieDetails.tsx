import _ from "lodash";
import { useEffect, useState } from "react";
import { useParams } from "react-router-dom";
import styled from "styled-components";
import { getMovies, Movie } from "../../api/movies";
import PageWrapper from "../../components/PageWrapper";

const Column1 = styled.div`
  width: 400px;
  margin: auto;
`;

const Column2 = styled.div`
  flex: 0 1 auto;
  flex-wrap: wrap;
`;

const MovieDetailsColumns = styled.div`
  display: flex;
  width: 100%;
  @media (max-width: 800px) {
    flex-direction: column;
  }
`;

const Poster = styled.img`
  width: auto;
  height: auto;
  max-height: 500px;
  display: block;
  margin: auto;
  padding-right: 20px;
  @media (max-width: 800px) {
    padding-right: 0px;
    padding-bottom: 20px;
  }
`;

const Title = styled.div`
  padding-bottom: 30px;
  @media (max-width: 800px) {
    text-align: center;
  }
`;

const useMovieData = () => {
  const { id } = useParams<{ id: string }>();
  const [movieData, setMovieData] = useState<Movie>();
  useEffect(() => {
    const retriveMovie = async () => {
      const result = await getMovies("");
      const movieDataFromResult = result.find((movie) => movie.id === id);
      setMovieData(movieDataFromResult);
    };
    retriveMovie();
  }, [id]);
  return movieData;
};

const MovieDetails: React.FC = () => {
  const movieData = useMovieData();

  if (!movieData) {
    return null;
  } else {
    return (
      <PageWrapper>
        <MovieDetailsColumns>
          <Column1>
            <Poster src={movieData.poster} alt="Poster" />
          </Column1>
          <Column2>
            <Title>
              MovieDetails: {movieData.title} ({movieData.classification})
            </Title>
            <div>
              <span>
                {new Date(movieData.released_on).getFullYear()} |{" "}
                {movieData.length} |{" "}
                {_.isArray(movieData.director)
                  ? movieData.director.join(", ")
                  : movieData.director}
              </span>
            </div>
            <div>
              <span>Cast: {movieData.cast.join(", ")}</span>
            </div>
            <div>
              <p>Movie Description: {movieData.overview}</p>
            </div>
          </Column2>
        </MovieDetailsColumns>
      </PageWrapper>
    );
  }
};

export default MovieDetails;
