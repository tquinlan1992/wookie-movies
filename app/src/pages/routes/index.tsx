import React from "react";
import { Route, Switch } from "react-router-dom";
import Home from "../Home";
import MovieDetails from "../movies/MovieDetails";
import { paths } from "./paths";

const routes: { path: string; component: JSX.Element }[] = [
  { path: paths.HOME, component: <Home /> },
  { path: paths.MOVIE_DETAILS, component: <MovieDetails /> },
];
const NotFound: React.FC = () => <h1>Page Not Found</h1>;

const Routes = () => (
  <Switch>
    {routes.map(({ path, component }) => (
      <Route exact path={path} key={path}>
        {component}
      </Route>
    ))}
    <Route component={NotFound} />
  </Switch>
);

export default Routes;
