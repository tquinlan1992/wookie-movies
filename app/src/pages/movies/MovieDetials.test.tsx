import { configure, mount } from "enzyme";
import MovieDetails from "./MovieDetails";
import * as moviesApi from "../../api/movies";
import toJson from "enzyme-to-json";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import { act } from "react-dom/test-utils";
import { Movie } from "../../api/movies";
import { MemoryRouter, Route } from "react-router-dom";
import { paths, toPaths } from "../routes/paths";

configure({ adapter: new Adapter() });

const waitForComponentToPaint = async (wrapper: any) => {
  await act(async () => {
    await new Promise((resolve) => setTimeout(resolve, 0));
    wrapper.update();
  });
};

function partialAs<Type>(input: Partial<Type>) {
  return input as Type;
}

test("render the  home  page", async () => {
  const movieDetailsToRender = partialAs<Movie>({
    title: "movie2",
    id: "movie2Id",
    poster: "movie2Poster",
    classification: "movie2Classification",
    released_on: new Date("1999").toString(),
    length: "movie2Length",
    director: ["director1", "director2"],
    cast: ["cast1", "cast2", "cast3"],
    overview: "movie2Overview",
  });
  const getMoviesSpy = jest.spyOn(moviesApi, "getMovies").mockReturnValue(
    Promise.resolve([
      partialAs<Movie>({
        id: "movie1Id",
      }),
      movieDetailsToRender,
      partialAs<Movie>({
        id: "movie3Id",
      }),
    ])
  );
  const MemoryRouterWrapper: React.FC = ({ children }) => (
    <MemoryRouter
      initialEntries={[
        toPaths.MOVIE_DETAILS({
          id: movieDetailsToRender.id,
          title: movieDetailsToRender.title,
        }),
      ]}
    >
      <Route path={paths.MOVIE_DETAILS}>{children}</Route>
    </MemoryRouter>
  );
  const wrapper = mount(
    <MemoryRouterWrapper>
      <MovieDetails />
    </MemoryRouterWrapper>
  );
  await waitForComponentToPaint(wrapper);
  expect(toJson(wrapper.find("MovieDetails"))).toMatchSnapshot();
  expect(getMoviesSpy).toHaveBeenCalled();
});
