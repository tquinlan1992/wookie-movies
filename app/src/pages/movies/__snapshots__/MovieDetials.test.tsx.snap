// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render the  home  page 1`] = `
<MovieDetails>
  <styled.div>
    <div
      className="sc-bdfBQB iNDsdm"
    >
      <styled.div>
        <div
          className="sc-hKgJUU hfhdcy"
        >
          <styled.div>
            <div
              className="sc-gsTEea ilesjl"
            >
              <styled.img
                alt="Poster"
                src="movie2Poster"
              >
                <img
                  alt="Poster"
                  className="sc-eCstlR ceXWvt"
                  src="movie2Poster"
                />
              </styled.img>
            </div>
          </styled.div>
          <styled.div>
            <div
              className="sc-dlfnuX kQmceR"
            >
              <styled.div>
                <div
                  className="sc-jSgvzq jwgnQ"
                >
                  MovieDetails: 
                  movie2
                   (
                  movie2Classification
                  )
                </div>
              </styled.div>
              <div>
                <span>
                  1998
                   |
                   
                  movie2Length
                   |
                   
                  director1, director2
                </span>
              </div>
              <div>
                <span>
                  Cast: 
                  cast1, cast2, cast3
                </span>
              </div>
              <div>
                <p>
                  Movie Description: 
                  movie2Overview
                </p>
              </div>
            </div>
          </styled.div>
        </div>
      </styled.div>
    </div>
  </styled.div>
</MovieDetails>
`;
